worlds/witness/test/__init__.py

```diff
@@ -0,0 +1,149 @@
+            event_location.access_rule = original_rules[event_location.name]
+
+    def assert_location_exists(self, location_name: str, strict_check: bool = True) -> None:
+        """
+        Assert that a location exists in this world.
+        If strict_check, also make sure that this (non-event) location COULD exist.
+        """
+
+        if strict_check:
+            self.assertIn(location_name, self.world.location_name_to_id, f"Location {location_name} can never exist")
+
+        try:
+            self.world.get_location(location_name)
+        except KeyError:
+            self.fail(f"Location {location_name} does not exist.")
+
+    def assert_location_does_not_exist(self, location_name: str, strict_check: bool = True) -> None:
+        """
+        Assert that a location exists in this world.
+        If strict_check, be explicit about whether the location could exist in the first place.
+        """
+
+        if strict_check:
+            self.assertIn(location_name, self.world.location_name_to_id, f"Location {location_name} can never exist")
+
+        self.assertRaises(
+            KeyError,
+            lambda _: self.world.get_location(location_name),
+            f"Location {location_name} exists, but is not supposed to.",
+        )
+
+    def assert_can_beat_with_minimally(self, required_item_counts: Mapping[str, int]) -> None:
+        """
+        Assert that the specified mapping of items is enough to beat the game,
+        and that having one less of any item would result in the game being unbeatable.
+        """
+        # Find the actual items
+        found_items = [item for item in self.multiworld.get_items() if item.name in required_item_counts]
+        actual_items: Dict[str, List[Item]] = {item_name: [] for item_name in required_item_counts}
+        for item in found_items:
+            if len(actual_items[item.name]) < required_item_counts[item.name]:
+                actual_items[item.name].append(item)
+
+        # Assert that enough items exist in the item pool to satisfy the specified required counts
+        for item_name, item_objects in actual_items.items():
+            self.assertEqual(
+                len(item_objects),
+                required_item_counts[item_name],
+                f"Couldn't find {required_item_counts[item_name]} copies of item {item_name} available in the pool, "
+                f"only found {len(item_objects)}",
+            )
+
+        # assert that multiworld is beatable with the items specified
+        self.assertTrue(
+            self.can_beat_game_with_items(item for items in actual_items.values() for item in items),
+            f"Could not beat game with items: {required_item_counts}",
+        )
+
+        # assert that one less copy of any item would result in the multiworld being unbeatable
+        for item_name, item_objects in actual_items.items():
+            with self.subTest(f"Verify cannot beat game with one less copy of {item_name}"):
```

Comment:
to get subtest to work with pytest, you just need to `pip install pytest-subtests`, however having a subtest inside an assertion feels very wrong to me.

Reply:
Oof, I was recommended this by Kaito and he managed to convince me. It *is* nice because I get *all* things that fail it at once rather than having to fix, run, fix, run, fix.

Reply:
Do you have a recommendation? Like, should I just get rid of the subtests in that function again?

Reply:
Not really, it is just odd naming though, so not a blocker.